postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,191 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+import logging
+
+import psycopg2
+import psycopg2.extras
+
+from datadog_checks.base.utils.db.sql import compute_sql_signature
+from datadog_checks.base.utils.db.statement_metrics import StatementMetrics, apply_row_limits
+
+try:
+    import datadog_agent
+except ImportError:
+    from ..stubs import datadog_agent
+
+
+logger = logging.getLogger(__name__)
+
+
+STATEMENTS_QUERY = """
+SELECT {cols}
+  FROM {pg_stat_statements_view} as pg_stat_statements
+  LEFT JOIN pg_roles
+         ON pg_stat_statements.userid = pg_roles.oid
+  LEFT JOIN pg_database
+         ON pg_stat_statements.dbid = pg_database.oid
+  WHERE pg_database.datname = %s
+  AND query != '<insufficient privilege>'
+"""
+
+# Required columns for the check to run
+PG_STAT_STATEMENTS_REQUIRED_COLUMNS = frozenset({'calls', 'query', 'total_time', 'rows'})
+
+PG_STAT_STATEMENTS_OPTIONAL_COLUMNS = frozenset({'queryid'})
+
+# Monotonically increasing count columns to be converted to metrics
+PG_STAT_STATEMENTS_METRIC_COLUMNS = {
+    'calls': 'postgresql.queries.count',
+    'total_time': 'postgresql.queries.time',
+    'rows': 'postgresql.queries.rows',
+    'shared_blks_hit': 'postgresql.queries.shared_blks_hit',
+    'shared_blks_read': 'postgresql.queries.shared_blks_read',
+    'shared_blks_dirtied': 'postgresql.queries.shared_blks_dirtied',
+    'shared_blks_written': 'postgresql.queries.shared_blks_written',
+    'local_blks_hit': 'postgresql.queries.local_blks_hit',
+    'local_blks_read': 'postgresql.queries.local_blks_read',
+    'local_blks_dirtied': 'postgresql.queries.local_blks_dirtied',
+    'local_blks_written': 'postgresql.queries.local_blks_written',
+    'temp_blks_read': 'postgresql.queries.temp_blks_read',
+    'temp_blks_written': 'postgresql.queries.temp_blks_written',
+}
+
+# Columns to apply as tags
+PG_STAT_STATEMENTS_TAG_COLUMNS = {
+    'datname': 'db',
+    'rolname': 'user',
+    'query': 'query',
+}
+
+DEFAULT_METRIC_LIMITS = {k: (100000, 100000) for k in PG_STAT_STATEMENTS_METRIC_COLUMNS.keys()}
+
+
+class PostgresStatementMetrics(object):
+    """Collects telemetry for SQL statements"""
+
+    def __init__(self, config):
+        self.config = config
+        # Cache results of monotonic pg_stat_statements to compare to previous collection
+        self._state = StatementMetrics()
+
+        # Available columns will be queried once and cached as the source of truth.
+        self._pg_stat_statements_columns = None
+        self._pg_stat_statements_query_columns = None
+
+    def _execute_query(self, cursor, query):
+        try:
+            cursor.execute(query)
+            return cursor.fetchall()
+        except (psycopg2.ProgrammingError, psycopg2.errors.QueryCanceled) as e:
+            logger.warning('Statement-level metrics are unavailable: %s', e)
+            return []
+
+    def init_pg_stat_statements_columns(self, db):
```

Comment:
Should this method be private? (Don't think it's meant to be used outside this helper class.)

```suggestion
    def _init_pg_stat_statements_columns(self, db):
```